### PR TITLE
fix: Fix signer to allow streaming unsigned payload without checksum & add integration test

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/1660

## Description of changes

Refer to comments on GH diff for details

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import Smithy
```
Comment (sichanyoo) [inline IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift:8]: Adds an integration test for streaming payload with unsigned chunks.

File: Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift [hunk 2]
```diff
@@ -256,12 +256,14 @@ public class AWSSigV4Signer: SmithyHTTPAuthAPI.Signer {
             }
         }
```
Comment (sichanyoo) [inline Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift:258]: By this point in this method, we know that the stream is going to be sent to server using chunked encoding. Therefore, regardless of checksum presence, if body is unsigned body, we have to resolve to `.streamingUnsignedPayloadTrailer`.

**Review by sichanyoo — COMMENTED** [review]
Comments for reviewers

File: IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift [hunk 3]
```diff
@@ -81,6 +82,46 @@ final class S3FlexibleChecksumsTests: S3XCTestCase {
...
+        func modifyBeforeTransmit(context: some MutableRequest<InputType, RequestType>) async throws {
+            context.getRequest().withHeader(name: "x-amz-content-sha256", value: "STREAMING-UNSIGNED-PAYLOAD-TRAILER")
+        }
+    }
+
+    class DisablePayloadSigningProvider: HttpInterceptorProvider {
+      func create<InputType, OutputType>() -> any HttpInterceptor<InputType, OutputType> {
+        return DisablePayloadSigning()
+      }
+    }
```
Comment (jbelkins) [inline IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift:99]: check indentation above

File: Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift [hunk 4]
```diff
@@ -256,12 +256,14 @@ public class AWSSigV4Signer: SmithyHTTPAuthAPI.Signer {
...
 
+        // STREAMING-UNSIGNED-PAYLOAD-TRAILER
+        if isUnsignedBody { return .streamingUnsignedPayloadTrailer }
+
         // streaming + eligible for chunked transfer
         if !checksumIsPresent {
```
Comment (jbelkins) [inline Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift:263]: should we reverse this `if-else`?  Weird to have a negated condition for an `if-else`

File: IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift [hunk 5]
```diff
@@ -81,6 +82,46 @@ final class S3FlexibleChecksumsTests: S3XCTestCase {
         XCTAssertEqual(data, originalData)
     }
 
+    // Streaming unsigned payload without checksum (chunked encoding)
+    class DisablePayloadSigning<InputType, OutputType>: HttpInterceptor {
```
Comment (jbelkins) [inline IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift:86]: Should this interceptor just be something we provide to customers?

**Review by jbelkins — COMMENTED** [review]
A couple minor fixes & questions

File: Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift (hunk 4 above)
Comment (jbelkins) [reply Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift:263]: this whole if-else can now break down to one statement BTW

File: IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift (hunk 5 above)
Comment (sichanyoo) [reply IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift:86]: We could, as part of some new enum that contains common interceptors.

File: Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift (hunk 4 above)
Comment (sichanyoo) [reply Sources/Core/AWSSDKHTTPAuth/Sources/AWSSDKHTTPAuth/AWSSigV4Signer.swift:263]: Changed as suggested.

File: IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift (hunk 3 above)
Comment (sichanyoo) [reply IntegrationTests/Services/AWSS3IntegrationTests/S3FlexibleChecksumsTests.swift:99]: Fixed indentation.

**Review by jbelkins — APPROVED** [review]

**Pull request merged**